lib/new.js

```diff
@@ -0,0 +1,687 @@
+/**
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License');
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const path = require('path');
+const puppeteer = require('puppeteer-core');const findChrome = require('./find_chrome');
+const {rpc} = require('../rpc');
+const debugApp = require('debug')('carlo:app');
+const debugServer = require('debug')('carlo:server');
+const {Color} = require('./color');
+const {HttpRequest} = require('./http_request');
+
+const fs = require('fs');
+const util = require('util');
+const {URL} = require('url');
+const EventEmitter = require('events');
+const fsReadFile = util.promisify(fs.readFile);
+
+let testMode = false;
+
+class App extends EventEmitter {
+  /**
+   * @param {!Puppeteer.Browser} browser Puppeteer browser
+   * @param {!Object} options
+   */
+  constructor(browser, options) {
+    super();
+    this.browser_ = browser;
+    this.options_ = options;    this.windows_ = new Map();    this.exposedFunctions_ = [];
+    this.pendingWindows_ = new Map();
+    this.windowSeq_ = 0;
+    this.www_ = [];
+  }
+
+  async init_() {
+    debugApp('Configuring browser');
+    let page;
+    await Promise.all([
+      this.browser_.target().createCDPSession().then(session => {
+        this.session_ = session;
+        if (this.options_.icon)
+          this.setIcon(this.options_.icon);
+      }),
+      this.browser_.defaultBrowserContext().
+          overridePermissions('https://domain', [
```

Comment:
format: style mismatch:
```javascript
58|      this.browser_.defaultBrowserContext().
59|          overridePermissions('https://domain', [
60|            'geolocation',
```
```suggestion
          overridePermissions('https://domain', ['geolocation','midi','notifications','camera','microphone','clipboard-read','clipboard-write']),
```

⏎␣⁺␣⁺ at column 50 should be removed.
Triggered rule # 182
```
  +1.roles not in {ARGUMENT}
	∧ +1.internal_type not in {StringLiteral}
	∧ +2.length ≤ 6
	∧ +2.roles not in {RIGHT}
	∧ +3.roles not in {NUMBER}
	∧ +4.reserved not in {,}
	∧ +5.reserved not in {(}
	∧ ^1.roles in {OR, OPERATOR} and not in {DECLARATION}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, (, .}
	∧ -2.diff_line = 0
	∧ -4.reserved not in {;}
	∧ -4.diff_offset ≤ 6
	⇒ y = ␣
	Confidence: 0.938. Support: 8.
```
⏎ at column 27 should be removed.
Triggered rule # 274
```
  +1.roles not in {ARGUMENT, CALLEE}
	∧ +1.reserved not in {=, {}
	∧ +2.roles not in {RIGHT}
	∧ +2.reserved = (
	∧ ^1.roles in {BLOCK} and not in {IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.roles not in {CALL}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 1
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	⇒ y = ⏎
	Confidence: 0.837. Support: 46.
```
⏎ at column 20 should be removed.
Triggered rule # 278
```
  +1.roles not in {ARGUMENT, CALLEE}
	∧ +1.reserved not in {=, {}
	∧ +2.length ≥ 3
	∧ +2.reserved not in {(}
	∧ +2.roles not in {RIGHT}
	∧ +3.roles in {EXPRESSION}
	∧ +5.internal_type not in {Identifier}
	∧ ^1.roles in {BLOCK} and not in {IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.roles not in {CALL}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 1
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	⇒ y = ∅
	Confidence: 0.812. Support: 8.
```
⏎ at column 29 should be removed.
Triggered rule # 176
```
  +1.roles not in {ARGUMENT}
	∧ +1.internal_type not in {StringLiteral}
	∧ +2.roles not in {RIGHT}
	∧ ^1.roles in {OPERATOR} and not in {DECLARATION}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved = ( and not in {,, :, .}
	∧ -2.diff_line = 0
	∧ -4.reserved not in {;}
	⇒ y = ∅
	Confidence: 0.979. Support: 24.
```
⏎ at column 22 should be removed.
Triggered rule # 288
```
  +1.roles not in {ARGUMENT, CALLEE}
	∧ +1.reserved not in {=, {}
	∧ +2.length ≤ 2
	∧ +2.reserved not in {(}
	∧ +2.roles not in {RIGHT}
	∧ ^1.roles in {BLOCK} and not in {IDENTIFIER, OPERATOR, DECLARATION}
	∧ ^2.roles not in {CALL}
	∧ -1.roles not in {STRING, EXPRESSION}
	∧ -1.reserved not in {,, :, ), .}
	∧ -1.diff_offset ≤ 1
	∧ -2.diff_line = 0
	∧ -2.label not in {<-space>}
	∧ -2.diff_col ≤ 3
	∧ -3.label not in {<newline>}
	∧ -4.roles not in {CALL}
	∧ -4.reserved not in {;}
	∧ •••start_col ≤ 17
	⇒ y = ∅
	Confidence: 0.815. Support: 100.
```
⏎ at column 26 should be removed.
Triggered rule # <N/A>
```
  +2.roles in {RIGHT}
	⇒ y = ␣
	Confidence: 0.997. Support: 195.
```